fedimint-recurringd/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use fedimint_api_client::api::net::Connector;
+use fedimint_client::{Client, ClientHandleArc, ClientModuleInstance};
+use fedimint_core::config::FederationId;
+use fedimint_core::core::{ModuleKind, OperationId};
+use fedimint_core::db::{Database, IDatabaseTransactionOpsCoreTyped, IRawDatabase};
+use fedimint_core::invite_code::InviteCode;
+use fedimint_core::secp256k1::hashes::sha256;
+use fedimint_core::util::SafeUrl;
+use fedimint_core::{Amount, BitcoinHash};
+use fedimint_derive_secret::DerivableSecret;
+use fedimint_ln_client::recurring::{
+    PaymentCodeId, PaymentCodeRootKey, RecurringPaymentError, RecurringPaymentProtocol,
+};
+use fedimint_ln_client::{LightningClientInit, LightningClientModule, LnReceiveState};
+use fedimint_mint_client::MintClientInit;
+use futures::StreamExt;
+use lightning_invoice::{Bolt11Invoice, Bolt11InvoiceDescription, Sha256};
+use lnurl::lnurl::LnUrl;
+use lnurl::pay::{LnURLPayInvoice, PayResponse};
+use lnurl::Tag;
+use tokio::sync::{Notify, RwLock};
+use tracing::{info, warn};
+
+use crate::db::{
+    load_federation_client_databases, open_client_db, try_add_federation_database,
+    FederationDbPrefix, PaymentCodeEntry, PaymentCodeInvoiceEntry, PaymentCodeInvoiceKey,
+    PaymentCodeKey, PaymentCodeNextInvoiceIndexKey, PaymentCodeVariant,
+};
+
+mod db;
+
+#[derive(Clone)]
+pub struct RecurringInvoiceServer {
+    db: Database,
+    clients: Arc<RwLock<HashMap<FederationId, ClientHandleArc>>>,
+    invoice_generated: Arc<Notify>,
+    base_url: SafeUrl,
+}
+
+impl RecurringInvoiceServer {
+    pub async fn new(db: impl IRawDatabase + 'static, base_url: SafeUrl) -> anyhow::Result<Self> {
+        let db = Database::new(db, Default::default());
+
+        let mut clients = HashMap::<_, ClientHandleArc>::new();
+
+        for (federation_id, db) in load_federation_client_databases(&db).await {
+            let mut client_builder = Client::builder(db).await?;
+            client_builder.with_module(LightningClientInit::default());
+            client_builder.with_module(MintClientInit);
```

Comment:
Should the dummy client be used here? I don't think `recurringd` should be doing anything with the primary module and it would eliminate a dependency on the mint client

Reply:
I don't think that would work since the federations don't have the dummy module in their config, so the client would detect it initialized without a primary module.